src/domain/dataverse/aggregate.ts

```diff
@@ -0,0 +1,176 @@
+  query: {
+    hasNext: boolean
+    isLoading: boolean
+    error: DataverseQueryError
+    filters: DataverseQueryFilters
+    setFilters: (newFilter: DataverseQueryFilters) => void
+    setLanguage: (newLng: string) => void
+  }
+  dataverse: DataverseDTO
+  loadDataverse: () => T.Task<void> | void
+}
+
+type ImmerSetReturnType = Partial<DataverseAggregate>
+
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+export const dataverseAggregate = (gateway: DataversePort) =>
+  createStore<DataverseAggregate, [['zustand/immer', never]]>(
+    immer<DataverseAggregate>((set, get) => ({
+      dataverse: [],
+      query: {
+        limit: 21,
+        hasNext: false,
+        isLoading: false,
+        error: null,
+        filters: { byType: 'all' },
+        language: 'en',
+        setIsLoading: (isLoading: boolean): void =>
+          set(
+            (state): ImmerSetReturnType => ({
+              query: { ...state.query, isLoading }
+            })
+          ),
+        setHasNext: (hasNext: boolean): void =>
+          set(
+            (state): ImmerSetReturnType => ({
+              query: { ...state.query, hasNext }
+            })
+          ),
+        setError: (error: DataverseQueryError): void =>
+          set(
+            (state): ImmerSetReturnType => ({
+              query: { ...state.query, error }
+            })
+          ),
+        setFilters: (newFilters: DataverseQueryFilters): void => {
+          const { clearDataverse } = get()
+          clearDataverse()
+          return set(
+            (state): ImmerSetReturnType => ({
+              query: { ...state.query, filters: { ...state.query.filters, ...newFilters } }
+            })
+          )
+        },
+        setLanguage: (newLng: string): void => {
+          const {
+            clearDataverse,
+            query: { setError }
+          } = get()
+          if (!newLng.length) {
+            return setError(
+              new Error(
+                `Oops.. An error occurred in <setLanguage> call.. Parameter cannot be empty: <${newLng}>..`
+              )
+            )
+          }
+          clearDataverse()
+          return set(
+            (state): ImmerSetReturnType => ({
+              query: { ...state.query, language: newLng }
+            })
+          )
+        }
+      },
+      clearDataverse: (): void =>
+        set(
+          (): ImmerSetReturnType => ({
+            dataverse: []
+          })
+        ),
+      addToDataverse: (newDataverse: DataverseEntity): void =>
+        set(
+          (state): ImmerSetReturnType => ({
+            dataverse: [...state.dataverse, ...newDataverse]
+          })
+        ),
+      loadDataverse: (): T.Task<void> | void => {
+        const { addToDataverse, query, dataverse } = get()
+        const { setIsLoading, setHasNext, setError, limit, language, filters } = query
+        const offset = dataverse.length
+
+        setIsLoading(true)
+        setError(null)
```

Comment:
I didn't catched that. You must avoid side effects in this function, as it is pure. Instead, you should place these impure statements in the `Task` monad.

Reply:
@ccamel 
I'm not sure to understand, we are already here in a TE monad, using matching pattern to execute side effects.
What do you mean by _ you should place these impure statements in the Task monad._ ?

Reply:
@fredericvilcot 

It's mainly the question of effects. If you call your function, it shouldn't produce any effect, it should just return the value `TaskEither`. This is not the case in your function, because if you call the function, you set the state to isLoading to true. The side effect should be inside the TE.

```ts
loadDataverse: (): T.Task<void> | void => {
        const { addToDataverse, query, dataverse } = get()
        const { setIsLoading, setHasNext, setError, limit, language, filters } = query
        const offset = dataverse.length

        setIsLoading(true) // <- side effect outside the TE
        setError(null) // <- side effect outside the TE

        return TE...
```

Reply:
@ccamel 
Sorry, I thought you were speaking about the `set...` functions inside the `TE.match` function.
I didn't noticed you were referring to these lines...
100% agree, I will wrap these side effects into a function that's called inside the `pipe`.